### ISSUE TITLE
zypper migration fails on get_zypper_pid()

### ISSUE THREAD

**Issue by marki555** [SUSE-Enceladus/cloud-regionsrv-client#198]
I have upgraded VM on Azure from SLES12SP5 to SLES15SP1 and then run zypper migration to SLES15SP3 without errors.
But when trying zypper migration (from 15SP3) to 15SP5, it failed for each repository with this error in /var/log/zypper.log:

```
2024-08-20 11:07:54 <1> sles12sp5(3488) [zypp::plugin++] PluginScript.cc(open):193 Open /usr/lib/zypp/plugins/urlresolver/susecloud()
2024-08-20 11:07:54 <1> sles12sp5(3488) [zypp::exec++] forkspawnengine.cc(start):181 Executing '/usr/lib/zypp/plugins/urlresolver/susecloud'
2024-08-20 11:07:54 <1> sles12sp5(3488) [zypp::exec++] forkspawnengine.cc(start):333 pid 3557 launched
2024-08-20 11:07:54 <1> sles12sp5(3488) [zypp::plugin++] PluginScript.cc(open):222 PluginScript[3557] /usr/lib/zypp/plugins/urlresolver/susecloud()
2024-08-20 11:07:54 <1> sles12sp5(3488) [zypp::plugin++] PluginScript.cc(send):272 PluginScript[3557] /usr/lib/zypp/plugins/urlresolver/susecloud ->send PluginFrame[RESOLVEURL](2){0}
2024-08-20 11:07:54 <2> sles12sp5(3488) [zypp::plugin] PluginScript.cc(receive):380 Unexpected EOF
2024-08-20 11:07:54 <5> sles12sp5(3488) [zypp-core] Exception.cc(log):186 PluginScript.cc(receive):381 THROW:    Receive: script died unexpectedly
2024-08-20 11:07:54 <5> sles12sp5(3488) [zypp-core] Exception.cc(log):186 History:
2024-08-20 11:07:54 <5> sles12sp5(3488) [zypp-core] Exception.cc(log):186  - [11-Resource temporarily unavailable]
2024-08-20 11:07:54 <5> sles12sp5(3488) [zypp-core] Exception.cc(log):186
2024-08-20 11:07:54 <2> sles12sp5(3488) [PLUGIN] PluginScript.cc(~PluginDumpStderr):75 ! Traceback (most recent call last):
2024-08-20 11:07:54 <2> sles12sp5(3488) [PLUGIN] PluginScript.cc(~PluginDumpStderr):75 !   File "/usr/lib/zypp/plugins/urlresolver/susecloud", line 139, in <module>
2024-08-20 11:07:54 <2> sles12sp5(3488) [PLUGIN] PluginScript.cc(~PluginDumpStderr):75 !     plugin.main()
2024-08-20 11:07:54 <2> sles12sp5(3488) [PLUGIN] PluginScript.cc(~PluginDumpStderr):75 !   File "/usr/lib/python3.6/site-packages/zypp_plugin.py", line 143, in main
2024-08-20 11:07:54 <2> sles12sp5(3488) [PLUGIN] PluginScript.cc(~PluginDumpStderr):75 !     self.__collect_frame()
2024-08-20 11:07:54 <2> sles12sp5(3488) [PLUGIN] PluginScript.cc(~PluginDumpStderr):75 !   File "/usr/lib/python3.6/site-packages/zypp_plugin.py", line 115, in __collect_frame
2024-08-20 11:07:54 <2> sles12sp5(3488) [PLUGIN] PluginScript.cc(~PluginDumpStderr):75 !     method(frame.headers, frame.body)
2024-08-20 11:07:54 <2> sles12sp5(3488) [PLUGIN] PluginScript.cc(~PluginDumpStderr):75 !   File "/usr/lib/zypp/plugins/urlresolver/susecloud", line 45, in RESOLVEURL
2024-08-20 11:07:54 <2> sles12sp5(3488) [PLUGIN] PluginScript.cc(~PluginDumpStderr):75 !     utils.credentials_files_are_equal(repo_credentials)
2024-08-20 11:07:54 <2> sles12sp5(3488) [PLUGIN] PluginScript.cc(~PluginDumpStderr):75 !   File "/usr/lib/python3.6/site-packages/cloudregister/registerutils.py", line 185, in credentials_files_are_equal
2024-08-20 11:07:54 <2> sles12sp5(3488) [PLUGIN] PluginScript.cc(~PluginDumpStderr):75 !     target_root = get_zypper_target_root()
2024-08-20 11:07:54 <2> sles12sp5(3488) [PLUGIN] PluginScript.cc(~PluginDumpStderr):75 !   File "/usr/lib/python3.6/site-packages/cloudregister/registerutils.py", line 1437, in get_zypper_target_root
2024-08-20 11:07:54 <2> sles12sp5(3488) [PLUGIN] PluginScript.cc(~PluginDumpStderr):75 !     zypper_cmd = get_zypper_command()
2024-08-20 11:07:54 <2> sles12sp5(3488) [PLUGIN] PluginScript.cc(~PluginDumpStderr):75 !   File "/usr/lib/python3.6/site-packages/cloudregister/registerutils.py", line 1400, in get_zypper_command
2024-08-20 11:07:54 <2> sles12sp5(3488) [PLUGIN] PluginScript.cc(~PluginDumpStderr):75 !     with open('/proc/%s/cmdline' % zypper_pid, 'r') as zypper_pid_file:
2024-08-20 11:07:54 <2> sles12sp5(3488) [PLUGIN] PluginScript.cc(~PluginDumpStderr):75 ! FileNotFoundError: [Errno 2] No such file or directory: '/proc/2003\n 3488/cmdline'
2024-08-20 11:07:54 <2> sles12sp5(3488) [zypp::exec] abstractspawnengine.cc(checkStatus):182 Pid 3557 exited with status 1
2024-08-20 11:07:54 <1> sles12sp5(3488) [zypp::plugin++] PluginScript.cc(close):229 Close:PluginScript[-1] /usr/lib/zypp/plugins/urlresolver/susecloud
2024-08-20 11:07:54 <1> sles12sp5(3488) [zypp::plugin++] PluginScript.cc(send):272 PluginScript[-1] /usr/lib/zypp/plugins/urlresolver/susecloud ->send PluginFrame[_DISCONNECT](0){0}
2024-08-20 11:07:54 <3> sles12sp5(3488) [zypp::plugin] PluginScript.cc(send):326 write(): [32-Broken pipe]

```

The reason is there are 2 `Zypp-main` processes and function `get_zypper_pid()` returns both PIDs separated with newline instead of only one. and then `get_zypper_command()` tries to open file `/proc/2003\n 3488/cmdline`.
These functions are in file "/usr/lib/python3.6/site-packages/cloudregister/registerutils.py", which belongs to package cloud-regionsrv-client-10.3.0-150300.13.3.1.x86_64. When I replaced line 1415 with `pid = pidData[0].strip().decode().partition('\n')[0]`, it was able to refresh repositories and zypper migration was able to run. There was no error with the repositories as other zypper commands (like zypper ref) worked fine when invoked from command-line, but not when invoked from within zypper migration.

Not sure if this is just the edge-case and why it hit me, but the code doesn't seem right as it breaks when the `ps` command finds more than 1 process.

**Comment by thimslugga** [comment]
I was able to workaround this issue by running `/usr/lib/zypper/commands/zypper-migration` to initiate the migration to newer service pack.

**Comment by marki555** [comment]
> I was able to workaround this issue by running `/usr/lib/zypper/commands/zypper-migration` to initiate the migration to newer service pack.

You mean running `zypper-migration` command instead of `zypper migration`? Yes, that could probably help. I wonder if nobody else was affected by this bug, because the code seems to be there for quite a long time. Or maybe the way the migration process is invoked changed, and that affected the `ps` output?

**Comment by rjschwei** [comment]
Resolved with 518cd8563e
